### PR TITLE
DockerCPURuntime

### PR DESCRIPTION
**Pull request by Wiezzel**

### CODE REVIEW

**Comment by Wiezzel** [comment]
Don't mind the Linux tests failing, it's most probably unrelated.

File: golem/envs/docker/cpu.py [hunk 1]
```diff
@@ -50,17 +56,227 @@ def from_dict(dict_: Dict[str, Any]) -> 'DockerCPUConfig':
...
+            image=image,
+            volumes=volumes,
+            command=command,
+            user=payload.user,
+            environment=payload.env,
+            working_dir=payload.work_dir,
+            host_config=host_config
+        )
+
+    def _change_status(
```
Comment (mfranciszkiewicz) [inline golem/envs/docker/cpu.py]: This method looks generic enough to be moved to e.g. parent class

File: golem/envs/docker/cpu.py [hunk 2]
```diff
@@ -50,17 +56,227 @@ def from_dict(dict_: Dict[str, Any]) -> 'DockerCPUConfig':
...
+            from_status = [from_status]
+
+        with self._status_lock:
+            if self._status not in from_status:
+                exp_status = "or ".join(map(str, from_status))
+                raise ValueError(
+                    f"Invalid status: {self._status}. Expected: {exp_status}")
+            self._status = to_status
+
+    def _wrap_status_change(
```
Comment (mfranciszkiewicz) [inline golem/envs/docker/cpu.py]: Same as above

File: golem/envs/docker/cpu.py [hunk 3]
```diff
@@ -50,17 +56,227 @@ def from_dict(dict_: Dict[str, Any]) -> 'DockerCPUConfig':
...
+            to_status: RuntimeStatus) -> None:
+        """ Assert that current Runtime status is the given one and change to
+            another one. Using lock to ensure atomicity. """
+
+        if isinstance(from_status, RuntimeStatus):
+            from_status = [from_status]
+
+        with self._status_lock:
+            if self._status not in from_status:
+                exp_status = "or ".join(map(str, from_status))
```
Comment (mfranciszkiewicz) [inline golem/envs/docker/cpu.py]:
```suggestion
                exp_status = " or ".join(map(str, from_status))
```

File: golem/envs/docker/cpu.py [hunk 4]
```diff
@@ -50,17 +56,227 @@ def from_dict(dict_: Dict[str, Any]) -> 'DockerCPUConfig':
...
+    CONTAINER_STOPPED: ClassVar[List[str]] = ["exited", "dead"]
+
+    STATUS_UPDATE_INTERVAL = 1.0  # seconds
+
+    def __init__(self, payload: DockerPayload, host_config: Dict[str, Any]) \
+            -> None:
+        image = f"{payload.image}:{payload.tag}"
+        volumes = [bind.target for bind in payload.binds]
+        if payload.command is not None:
+            command: Optional[List[str]] = [payload.command] + payload.args
```
Comment (mfranciszkiewicz) [inline golem/envs/docker/cpu.py]: Maybe `payload.command` should also contain arguments?

File: golem/envs/docker/cpu.py (hunk 1 above)
Comment (Wiezzel) [reply golem/envs/docker/cpu.py]: Done.

File: golem/envs/docker/cpu.py (hunk 2 above)
Comment (Wiezzel) [reply golem/envs/docker/cpu.py]: Done.

File: golem/envs/docker/cpu.py (hunk 4 above)
Comment (Wiezzel) [reply golem/envs/docker/cpu.py]: Done.

**Review by maaktweluit — APPROVED** [review]
LGTM!

**Pull request merged**